fetch.bs

```diff
@@ -3835,6 +3836,13 @@ these steps:
       </ul>
 
       <p>then return a <a>network error</a>.
+
+     <li>
+      <p>Set <var>actualResponse</var>'s <a for=response>URL</a>'s <a for=url>fragment</a> to
+      <var>request</var>'s <a for=request>current URL</a>'s <a for=url>fragment</a>.
+
+      <p class=note>There is currently no mechanism for the service worker to control
+      <var>actualResponse</var>'s <a for=response>URL</a>'s <a for=url>fragment</a>.
```

Comment:
If I'm understanding this right, it means:

* If a SW-less document contains `<img src="redirect.py?url=colors.svg%23green#red">`, we'll render `colors.svg#green`, since the location URL algorithm prefers the `Location` header's fragment.
* If that same document has a service worker which simply does `ev.respondWith(fetch(ev.request))`, the FetchEvent will be resolved with a Response containing `colors.svg#green` (since redirects were followed), but then this logic will replace the fragment with `#red`.

Am I understanding this right? Is that intentional?

That seems kinda weird, though I don't see clear alternatives given service worker caches should not be sensitive to fragments, and service workers take over URLs. One possibility, which is janky in a different way (but has a nice symmetry with redirects in general), would be to say:

* When a SW gets a FetchEvent, remove the fragment from the request URL. The SW, like an HTTP intermediary, should not get to react differently on request fragments.
* The response URL preserves the fragment. But, due to step 1, any fragments would come from redirects or the service worker itself. We haven't yet incorporated the request fragment.
* When we consume a SW-forwarded response, apply the request fragment similar to redirects: if the response URL had a fragment, use that. If not, use the request fragment.

(By the way, are there subresource contexts other than SVGs where the fragments matter?)

Reply:
And I guess both formulations have another weird corner: the response's URL isn't a field per se. It's a pointer to the last entry in the URL list. In which case, what does it mean to update it? Do you write into the last entry of the URL list but leave the others alone? What if the URL list is empty because this was script-constructed?

https://fetch.spec.whatwg.org/#concept-response-url

Reply:
I agree It would be nice if respondWith() functioned similar to redirects when it came to fragments.

Reply:
Thanks @davidben, you're right, this isn't great. I think your model is better but it would change `fetchEvent.request.url`. Is Chrome willing to experiment with that? I suppose synthetic responses would always end up with the fragment of the request from the document, which seems about right, but the current wording indeed isn't adequate.

Perhaps I should split out the service worker change into its own issue and pull request (it already has its own tests) as that clearly warrants some further consideration.

Reply:
(@wanderview is probably a better person to answer what we're willing to experiment with. I don't actually work on service worker stuff. I'm just really distractible and like to stick my nose into everything. Although I see I'm rapidly becoming "the redirect person" which is... a life decision. 😄)

I was pondering other ideas, and they broke on this case, so probably something to include in tests. Suppose the server responds to requests as follows:

GET /a => 302; Location: /b#foo
GET /b => 302; Location: /c
GET /c => 200

We'd want `fetch("/a")` to get a redirect path of /a -> /b#foo -> /c#foo. That implies that the fragment fixup logic needs to be part of `{redirect: "follow"}` and not something we only apply to the final response. However, `fetch("/a#foo")` would then go /a#foo -> /b#foo -> /c#foo and there isn't enough information to know whether c's fragment was intrinsic or came from the request fragment. That means any caching fetch intermediary must store the result of a fragment-less fetch. Otherwise you lose information and cannot correctly share cache entries between /a#foo and /a#bar.

Also, all this relies on it being impossible for the server to redirect and *remove* a fragment. (If the server sends a fragment-less location header, we fill in the request URL's fragment.) If it were possible, we'd need to store extra information in the response. That is, if the server says:

GET /a => 302; Location: /b#:magic-syntax-to-remove-fragment

Then we have /a -> /b, but also /a#foo -> /b, but that conflicts with the "When we consume a SW-forwarded response, [...]" logic. We'd need responses to store an extra copyFragmentFromRequestURL boolean.

To that end, that does suggest one more possible model which doesn't need to change FetchEvent:

* Responses have a "copy fragment from request URL" boolean property
* SW still sees the fragment and are allowed to fetch as before
* Fetch sets the new property to true if every Location header we followed was fragment-agnostic
* When you get a response from a SW, clobber the fragment if that property is true.

But this one seems even more fussy and probably isn't what we want. Yet another possibility would be to move the fragment fixup logic from FetchEvent.respondWith() to CacheStorage. Something like:

* Responses track whether fragments came from the request URL or not as above.
* When you serialize a response, don't serialize fragments that came from the request URL, only intrinsic ones.
* When you get a response out from CacheStorage.match(), apply the fragment fixup logic.